lib/fir/statements.h

```diff
@@ -0,0 +1,231 @@
+      BasicBlock *defaultBlock, const ValueSuccPairListType &args);
+  BasicBlock *defaultSucc() const;
+  std::list<BasicBlock *> succ_blocks() const override;
+  const Evaluation &getCond() const;
+
+private:
+  explicit SwitchStmt(const Evaluation &condition, BasicBlock *defaultBlock,
+      const ValueSuccPairListType &args);
+  Evaluation condition;
+  ValueSuccPairListType valueSuccPairs;
+};
+
+struct IndirectBrStmt : public TerminatorStmt_impl {
+  using TargetListType = std::vector<BasicBlock *>;
+  ~IndirectBrStmt() = default;
+  static IndirectBrStmt *Create(
+      Variable *variable, const TargetListType &potentialTargets);
+
+private:
+  explicit IndirectBrStmt(
+      Variable *variable, const TargetListType &potentialTargets)
+    : variable{variable}, potentialTargets{potentialTargets} {}
+  Variable *variable;
+  TargetListType potentialTargets;
+};
+
+struct UnreachableStmt : public TerminatorStmt_impl {
+  static UnreachableStmt *Create();
+  ~UnreachableStmt() = default;
+
+private:
+  explicit UnreachableStmt() = default;
+};
+
+struct ActionStmt_impl : public Stmt_impl {
+  using ActionTrait = std::true_type;
+  // TODO: DynamicType is a placeholder for now
+  std::optional<evaluate::DynamicType> type_{std::nullopt};
+};
+
+struct AssignmentStmt
+  : public ActionStmt_impl,
+    private SumTypeCopyMixin<std::variant<const parser::AssignmentStmt *,
+        const parser::AssignStmt *>> {
+  friend Statement;
+  ~AssignmentStmt() = default;
+  static AssignmentStmt *Create(const parser::AssignmentStmt *s);
+  static AssignmentStmt *Create(const parser::AssignStmt *s);
+
+private:
+  explicit AssignmentStmt(const parser::AssignmentStmt *s)
+    : SumTypeCopyMixin{s} {}
+  explicit AssignmentStmt(const parser::AssignStmt *s) : SumTypeCopyMixin{s} {}
+};
+
+struct PointerAssignStmt : public ActionStmt_impl {
+  friend Statement;
+  ~PointerAssignStmt() = default;
+  static PointerAssignStmt *Create(const parser::PointerAssignmentStmt *s);
+
+private:
+  explicit PointerAssignStmt(const parser::PointerAssignmentStmt *s)
+    : assign{s} {}
+  const parser::PointerAssignmentStmt *assign;
+};
+
+struct MemoryStmt_impl : public ActionStmt_impl {};
+struct LoadStmt : public MemoryStmt_impl {};
+struct StoreStmt : public MemoryStmt_impl {};
+
+struct AllocateStmt : public MemoryStmt_impl {
+  friend Statement;
+  ~AllocateStmt() = default;
+  static AllocateStmt *Create(const parser::AllocateStmt *s);
+
+private:
+  explicit AllocateStmt(const parser::AllocateStmt *s) : alloc{s} {}
+  const parser::AllocateStmt *alloc;
+};
+
+struct DeallocateStmt : public MemoryStmt_impl {
+  friend Statement;
+  ~DeallocateStmt() = default;
+  static DeallocateStmt *Create(const parser::DeallocateStmt *s);
+
+private:
+  explicit DeallocateStmt(const parser::DeallocateStmt *s) : dealloc{s} {}
+  const parser::DeallocateStmt *dealloc;
+};
+
+struct CastingStmt_impl : public ActionStmt_impl {};
```

Comment:
Data conversions will be explicit operations in RHS expressions.

Reply:
More vestigal stuff from LLVM.